### PR TITLE
spi: introduce `struct spi_dt_spec`

### PR DESCRIPTION
**Pull request by JordanYates**
Introduces the `struct spi_dt_spec` type, which contains the complete SPI bus information derived from devicetree. It serves the same purpose as `struct gpio_dt_spec` in that it can be constructed automatically in `DEVICE_DT_INST_DEFINE` macros and provided as a single handle to SPI API calls.

I believe the extension over `SPI_CS_CONTROL_PTR_DT` is justified as it further simplifies usage and allows holistic checking of bus readiness in `spi_is_ready`.

The spi_nor flash driver is converted as a demonstration of the boilerplate reduction that is enabled.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

### CODE REVIEW

File: include/drivers/spi.h [hunk 1]
```diff
@@ -363,6 +417,28 @@ __subsystem struct spi_driver_api {
...
+ * @param spec SPI specification from devicetree
+ *
+ * @retval true if the SPI bus is ready for use.
+ * @retval false if the SPI bus is not ready for use.
+ */
+static inline bool spi_is_ready(const struct spi_dt_spec *spec)
+{
+	/* Validate bus is ready */
+	if (!device_is_ready(spec->bus)) {
+		return -EINVAL;
```
Comment (mbolivar-nordic) [inline include/drivers/spi.h:432]: I think we usually use ENODEV for devices that are not ready.

File: drivers/flash/spi_nor.c [hunk 2]
```diff
@@ -1013,34 +1007,15 @@ static int setup_pages_layout(const struct device *dev)
...
 {
-	struct spi_nor_data *data = dev->data;
-	uint8_t jedec_id[SPI_NOR_MAX_ID_LEN];
 	const struct spi_nor_config *cfg = dev->config;
+	uint8_t jedec_id[SPI_NOR_MAX_ID_LEN];
 	int rc;
 
-	data->spi = device_get_binding(DT_INST_BUS_LABEL(0));
-	if (!data->spi) {
+	/* Validate bus is ready */
```
Comment (mbolivar-nordic) [inline drivers/flash/spi_nor.c:1014]: The comment is a bit misleading since spi_is_ready() also checks the chip select device.

**Review by mbolivar-nordic — COMMENTED** [review]
Seems reasonable to me. Couple of nitpicks.

File: include/drivers/spi.h (hunk 1 above)
Comment (JordanYates) [reply include/drivers/spi.h:432]: From a quick browse it looks like about a 50/50 split between ENODEV and EINVAL in the drivers.
A convention should probably be established in another PR.

**Review by henrikbrixandersen — APPROVED** [review]
Looks good.

File: include/drivers/spi.h [hunk 3]
```diff
@@ -297,6 +297,60 @@ struct spi_config {
...
+ * <tt>SPI_DT_SPEC_GET(DT_DRV_INST(inst), operation_, delay_)</tt>.
+ *
+ * This macro is not available in C++.
+ *
+ * @param inst Devicetree instance number
+ * @param operation_ the desired @p operation field in the struct spi_config
+ * @param delay_ the desired @p delay field in the struct spi_config's
+ *               spi_cs_control, if there is one
+ */
+#define SPI_DT_SPEC_GET_INST(inst, operation_, delay_) \
```
Comment (gmarull) [inline include/drivers/spi.h:350]: I think this API should follow the same naming convention useg, e.g. by GPIO (`GPIO_DT_SPEC_INST_GET` and not `GPIO_DT_SPEC_GET_INST`)

**Review by gmarull — DISMISSED** [review]

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
I really want to insist on ENODEV, sorry.

**Comment by JordanYates** [comment]
> I really want to insist on ENODEV, sorry.

That has already been updated.

**Comment by JordanYates** [comment]
>> I really want to insist on ENODEV, sorry.

> That has already been updated.

The function implementation was wrong, as I set the return type to be `bool` to match `device_is_ready` and then returned integer error codes. Interesting that the bad type wasn't caught at compile time.

**Comment by JordanYates** [comment]
@gmarull  @mbolivar-nordic I believe the concerns have been addressed

File: include/drivers/spi.h [hunk 4]
```diff
@@ -517,6 +557,24 @@ static inline int spi_write(const struct device *dev,
...
+ * This is equivalent to:
+ *
+ *     spi_write(spec->bus, &spec->config, tx_bufs);
+ *
+ * @param spec SPI specification from devicetree
+ * @param tx_bufs Buffer array where data to be sent originates from.
+ *
+ * @retval a value from spi_write()
+ */
+static inline int spi_write_dt(const struct spi_dt_spec *spec,
```
Comment (gmarull) [inline include/drivers/spi.h:572]: Do we expect to have `_dt` API helpers for other drivers as well? That is, should a developer expect to find, e.g., `gpio_pin_get_dt`? I think this is an important point for APIs consistency. In the case of GPIO, `_dt` API helpers are only available for configuration related APIs, but nothing else.

I don't have a strong opinion here, but I'd personally not mix DT with _operation_ APIs. Non-blocking comment, but I think it's worth discussing before merging these changes.

Comment (JordanYates) [reply include/drivers/spi.h:572]: > Do we expect to have _dt API helpers for other drivers as well?

I certainly hope so, writing boilerplate when you don't need to is annoying.
#36477 added the GPIO versions you mentioned :)

**Review by gmarull — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Pull request merged**